gb-apu/src/channel/volume_envelope.rs

```diff
@@ -0,0 +1,54 @@
+#[derive(PartialEq, Debug)]
+pub enum Direction {
+    Inc,
+    Dec,
+}
+#[derive(Debug)]
+pub struct VolumeEnvelope {
+    pub initial_volume: u8,
+    pub envelope_direction: Direction,
+    pub period: u8,
+    pub volume: u8,
+    counter: u8,
+}
+
+impl Default for VolumeEnvelope {
+    fn default() -> Self {
+        Self {
+            initial_volume: 0,
+            envelope_direction: Direction::Dec,
+            period: 0,
+            volume: 0,
+            counter: 0,
+        }
+    }
+}
+
+impl VolumeEnvelope {
+    pub fn step(&mut self) {
+        if self.period == 0 {
+            return;
+        }
+        if self.counter == 0 {
+            self.counter = self.period;
+            self.update_volume();
+        } else {
+            self.counter -= 1;
```

Comment:
Why we aren't taking the direction into account here ?

Reply:
The direction is used in the update_volume method

Reply:
I see, I was confusing `counter` and `volume`